objects/src/accounts/vault.rs

```diff
@@ -59,11 +65,161 @@ impl AccountVault {
+            current => {
+                let current: FungibleAsset =
+                    current.try_into().expect("tree only contains valid assets");
+                current.add(asset).map_err(AccountError::AddFungibleAssetBalanceError)?
+            }
+        };
+        self.asset_tree.insert(new.key().into(), new.into());
+
+        // return the new asset
+        Ok(new)
+    }
+
+    /// Add the specified non-fungible asset to the vault.
+    ///
+    /// # Errors
+    /// - If the vault already contains the same non-fungible asset.
+    fn add_non_fungible_asset(
+        &mut self,
+        asset: NonFungibleAsset,
+    ) -> Result<NonFungibleAsset, AccountError> {
+        // add non-fungible asset to the vault
+        let old = self.asset_tree.insert(asset.key().into(), asset.into());
+
+        // if the asset already exists, return an error
+        if old != [ZERO, ZERO, ZERO, ZERO] {
+            return Err(AccountError::DuplicateNonFungibleAsset(asset));
+        }
+
+        Ok(asset)
+    }
+
+    // REMOVE ASSET
+    // --------------------------------------------------------------------------------------------
+    /// Remove the specified asset from the vault.
+    ///
+    /// # Errors
+    /// - The fungible asset is not found in the vault.
+    /// - The amount of the fungible asset in the vault is less than the amount to be removed.
+    /// - The non-fungible asset is not found in the vault.
+    pub fn remove_asset(&mut self, asset: Asset) -> Result<Asset, AccountError> {
+        Ok(match asset {
+            Asset::Fungible(asset) => Asset::Fungible(self.remove_fungible_asset(asset)?),
+            Asset::NonFungible(asset) => Asset::NonFungible(self.remove_non_fungible_asset(asset)?),
+        })
+    }
+
+    /// Remove the specified fungible asset from the vault.
+    ///
+    /// # Errors
+    /// - The asset is not found in the vault.
+    /// - The amount of the asset in the vault is less than the amount to be removed.
+    fn remove_fungible_asset(
+        &mut self,
+        asset: FungibleAsset,
+    ) -> Result<FungibleAsset, AccountError> {
+        // fetch the asset from the vault.
+        let mut current: FungibleAsset = match self.asset_tree.get_value(asset.key().into()) {
+            current if current == [ZERO, ZERO, ZERO, ZERO] => {
+                return Err(AccountError::FungibleAssetNotFound(asset))
+            }
+            current => current.try_into().expect("tree only contains valid assets"),
+        };
+
+        // subtract the amount of the asset to be removed from the current amount.
+        current
+            .sub(asset.amount())
+            .map_err(AccountError::SubtractFungibleAssetBalanceError)?;
+
+        // if the amount of the asset is zero, remove the asset from the vault.
+        let new = match current.amount() {
+            0 => {
+                // return ZERO value to insert into the vault
+                [ZERO, ZERO, ZERO, ZERO]
+            }
```

Comment:
This won't work properly right now as `TieredSmt` doesn't handle deletions correctly.

Reply:
Will add a TODO